core/trino-main/src/main/java/io/trino/sql/planner/optimizations/StreamPropertyDerivations.java

```diff
@@ -342,6 +342,13 @@ public StreamProperties visitExchange(ExchangeNode node, List<StreamProperties>
                     if (node.getPartitioningScheme().getPartitioning().getHandle().equals(FIXED_ARBITRARY_DISTRIBUTION)) {
                         return new StreamProperties(FIXED, Optional.empty(), false);
                     }
+                    // if this is a transaction and there are no arguments we are safe to provide empty optional as there will be only one stream
+                    if (node.getPartitioningScheme().getPartitioning().getHandle().getTransactionHandle().isPresent()) {
```

Comment:
> if this is a transaction

do you mean transactional table?

> `node.getPartitioningScheme().getPartitioning().getHandle().getTransactionHandle().isPresent()`

this doesn't let us recognize what kind of table we're dealing with.
in fact, i'd expect this to be always true whenever we're dealing with connector-provided partitioning.
i am validating my understanding with `Remove redundant null-friendliness` commit in https://github.com/trinodb/trino/pull/10293

What about dropping this condition, and adding the "if arguments list is empty" logic directly to the code below?

Reply:
> do you mean transactional table?

this is exactly what I meant, was that an incorrect assumption ?

> this doesn't let us recognize what kind of table we're dealing with.

any idea how to recognize if we are dealing with transactional table ?

> What about dropping this condition, and adding the "if arguments list is empty" logic directly to the code below?

actually I wanted to avoid that, doing it this way will change the behaviour for all the situation and I want the change only for transactional tables. There is an explicit check: 
```
checkArgument(distribution == SINGLE || !this.partitioningColumns.equals(Optional.of(ImmutableList.of())),
                    "Multiple streams must not be partitioned on empty set");
```
modifying the logic you mentioned may cause this check not to fail for situation when it should. I just wanted to make it pass for transactional table as we know there will be one bucket and thus one stream only.

Reply:
Here is how it works

1. Hive provides bucketing function to be used when distributing writes. The function is 0-arg, because it's an artificial bucketing. The point is -- we want to have exactly one writer to the table.
2. `StreamPropertyDerivations` chokes on the 0-arg bucketing function. 

The fix can be

1. make `StreamPropertyDerivations` not choke on that (like you did)
2. find some other way for a connector to make sure there is only one writer
3. make Hive fool engine -- declare false argument to bucketing function, pretending it's not 0-arg
   - that would be working around engine's limitation. We shouldn't need to do that though
4. anything else? -- @electrum might know better


cc @losipiuk @arhimondr

Reply:
I just wonder if this is an accidentally choking or if it was made on purpose like this in which case removing that choking won't break some other cases.